pallets/confidential-asset/src/lib.rs

```diff
@@ -36,19 +51,91 @@ pub trait Trait:
 {
 }
 
+/// Wrapper for Ciphertexts that correspond to EncryptedAssetId.
+/// This is needed since `mercat::asset_proofs::elgamal_encryption::CipherText` implements
+/// Encode and Decode, instead of deriving them. As a result, the EncodeLike operator is
+/// not automatically implemented.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, VecU8StrongTyped, Default)]
+pub struct EncryptedAssetIdWrapper(pub Vec<u8>);
+
+/// Wrapper for Ciphertexts that correspond to EncryptedBalance.
+/// This is needed since `mercat::asset_proofs::elgamal_encryption::CipherText` implements
+/// Encode and Decode, instead of deriving them. As a result, the EncodeLike operator is
+/// not automatically implemented.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, VecU8StrongTyped, Default)]
+pub struct EncryptedBalanceWrapper(pub Vec<u8>);
+
+/// A mercat account consists of the public key that is used for encryption purposes and the
+/// encrypted asset id. The encrypted asset id also acts as the unique identifier of this
+/// struct.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, Default)]
+pub struct MercatAccount {
+    pub encrypted_asset_id: EncryptedAssetIdWrapper,
+    pub encryption_pub_key: EncryptionPubKey,
+}
+
 decl_storage! {
-    trait Store for Module<T: Trait> as Asset {
+    trait Store for Module<T: Trait> as ConfidentialAsset {
+
+        /// Contains the mercat accounts for an identity.
+        pub MercatAccounts get(fn mercat_account): double_map hasher(blake2_128_concat) IdentityId, hasher(blake2_128_concat) EncryptedAssetIdWrapper  => MercatAccount;
+
+        /// Contains the encrypted balance of a mercat account.
+        pub MercatAccountBalance get(fn mercat_account_balance):  double_map hasher(blake2_128_concat) IdentityId, hasher(blake2_128_concat) EncryptedAssetIdWrapper =>EncryptedAssetIdWrapper;
+
+        /// Contains the list of all valid ticker names.
+        /// The process around creating and storing this data will likely change as a result of CRYP-153.
+        pub ValidAssetIds get(fn valid_asset_ids): Vec<AssetId>;
     }
 }
 
 // Public interface for this runtime module.
 decl_module! {
     pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+
+        /// TODO: CRYP-153 will most likely change it into a private function that is called
+        /// when a new ticker name is registered.
+        #[weight = 1_000_000_000]
+        pub fn store_valid_asset_ids(
```

Comment:
@aafshar-poly  I really don't get it why we need this?

Reply:
MERCAT needs to maintain a list of valid `asset_id`s (aka ticker names). We are planning to share the `SecurityToken` struct between core and mercat, however we need to have a way of flagging assets as confidential vs non-confidential. We can:
1. Modify the `create_asset()` API to take in a flag for this.
2. or add another API to flag an asset as confidential, after it's created.

Reply:
Or have `create_asset()` make non-confidential assets, and another API to make confidential ones:
```
_create_asset(...,confidential: bool);
create_asset(...) {_create_asset(..., false)}
create_confidential_asset(...) {_create_asset(..., true)}
```

Reply:
Thanks, @ParnianAlimi, & @aafshar-poly to clarify. I would suggest having a different API if we have a different data structure for the confidential assets. 

For now, let's rename this dispatchable to `create_confidential_asset()` and have the same underlying functionality and add a TODO for the expected functionality.